infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java

```diff
@@ -0,0 +1,149 @@
+    Deployment baseDeployment = createEmptyDeployment(DEFAULT_DEPLOYMENT_NAME);
+    PodTemplateSpec basePodTemplate = baseDeployment.getSpec().getTemplate();
+    ObjectMeta basePodMeta = basePodTemplate.getMetadata();
+    PodSpec baseSpec = basePodTemplate.getSpec();
+
+    Set<String> containerNames = new HashSet<>();
+    Set<String> initContainerNames = new HashSet<>();
+    Set<String> volumes = new HashSet<>();
+    Set<String> pullSecrets = new HashSet<>();
+    for (PodData podData : podsData) {
+      // if there are entries with such keys then values will be overridden
+      ObjectMeta podMeta = podData.getMetadata();
+      basePodMeta.getLabels().putAll(podMeta.getLabels());
+      basePodMeta.getAnnotations().putAll(podMeta.getAnnotations());
+      basePodMeta.getAdditionalProperties().putAll(podMeta.getAdditionalProperties());
+
+      for (Container container : podData.getSpec().getContainers()) {
+        String containerName = container.getName();
+
+        // generate container name to avoid collisions
+        while (!containerNames.add(container.getName())) {
+          containerName = NameGenerator.generate(container.getName(), 4);
+          container.setName(containerName);
+        }
+
+        // store original recipe machine name
+        basePodMeta
+            .getAnnotations()
+            .put(
+                format(MACHINE_NAME_ANNOTATION_FMT, containerName),
+                Names.machineName(podMeta, container));
+
+        baseSpec.getContainers().add(container);
+      }
+
+      for (Container initContainer : podData.getSpec().getInitContainers()) {
+        // generate container name to avoid collisions
+        while (!initContainerNames.add(initContainer.getName())) {
+          initContainer.setName(NameGenerator.generate(initContainer.getName(), 4));
+        }
+        baseSpec.getInitContainers().add(initContainer);
+      }
+
+      for (Volume volume : podData.getSpec().getVolumes()) {
+        if (!volumes.add(volume.getName())) {
+          throw new ValidationException(
+              format(
+                  "Pods have to have volumes with unique names but there are multiple `%s` volumes",
+                  volume.getName()));
+        }
+        baseSpec.getVolumes().add(volume);
+      }
+
+      for (LocalObjectReference pullSecret : podData.getSpec().getImagePullSecrets()) {
+        if (pullSecrets.add(pullSecret.getName())) {
+          // add pull secret only if it is not present yet
+          baseSpec.getImagePullSecrets().add(pullSecret);
+        }
+      }
+
+      // if there are entries with such keys then values will be overridden
+      baseSpec.getAdditionalProperties().putAll(podData.getSpec().getAdditionalProperties());
+    }
+
+    Map<String, String> matchLabels = new HashMap<>();
+    matchLabels.put(DEPLOYMENT_NAME_LABEL, baseDeployment.getMetadata().getName());
+    basePodMeta.getLabels().putAll(matchLabels);
```

Comment:
I tested this PR manually and it worked just fine.
:smile_cat: 
It works because it is fixed here https://github.com/eclipse/che/blob/0f8471014130c72b06c40b2b35137d72f3756dd9/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java#L162.
Will fix it as well, thanks for catching.

Reply:
I've rechecked. And selector does not have to contain all entries from pod labels
![Screenshot_20190320_104916](https://user-images.githubusercontent.com/5887312/54671200-15d5ea00-4afe-11e9-92ed-c5392a111a2e.png)
@metlos Can you elaborate more, why do you think that this code does not ensure matching deployment selector to pod label?